### PR TITLE
feat(node, core): add `pay_program_rent` extrinsic/syscall

### LINKED ISSUE #2509: Ability to pay rent for programs
### Problem to Solve

after #2301 programs will be stopped at the end of free rent period. User should be able to pay additional rent to prolong the interval.

### Possible Solution

syscall and extrinsic

### Notes

_No response_

### PR DESCRIPTION
**Pull request by gshep**
Resolves #2509 

- extrinsic
- syscall
- test case that syscall works
- corresponding weights

@gear-tech/dev

### CODE REVIEW

File: core-errors/src/lib.rs [hunk 1]
```diff
@@ -220,6 +220,15 @@ pub enum ExecutionError {
     /// An error occurs in attempt to parse invalid string in `gr_debug` sys-call.
     #[display(fmt = "Invalid debug string passed in `gr_debug` sys-call")]
     InvalidDebugString = 0,
+
+    /// The error occurs when program's balance is less than rent it tries to pay.
+    #[display(fmt = "Existing value {value_left} is not enough to pay rent {rent}")]
+    NotEnoughValue {
```
Comment (DennisInSky) [inline core-errors/src/lib.rs:226]: Isn't it too general? Will it be better `NotEnoughValueForRent`?

File: core-backend/wasmi/src/funcs/mod.rs [hunk 2]
```diff
@@ -1279,6 +1279,31 @@ where
...
+                    let BlockNumberWithHash {
+                        hash: program_id,
+                        bn: block_count,
+                    } = ctx.read_as(read_bn_pid)?;
+
+                    let state = ctx.host_state_mut();
+                    state
+                        .ext
+                        .pay_rent(program_id.into(), block_count)
+                        .map_err(Into::into)
```
Comment (DennisInSky) [inline core-backend/wasmi/src/funcs/mod.rs:1300]: Have you ever had an idea of moving such pieces of code (essence of the call) into some common place as a func and call it from sandbox and wasmi implementations? Or it is not worth it at all?

File: core-errors/src/lib.rs (hunk 1 above)
Comment (gshep) [reply core-errors/src/lib.rs:226]: agree

File: core-backend/wasmi/src/funcs/mod.rs (hunk 2 above)
Comment (gshep) [reply core-backend/wasmi/src/funcs/mod.rs:1300]: Yep, repetition is not anything good. However if we would implement sandbox on our side wasmi backend-executor would also be removed. So not much sense for now.

**Comment by NikVolf** [comment]
How is it supposed to be used? User won't know in advance how much he is even paying.

**Comment by NikVolf** [comment]
Syscall should take some balance, not block count

I can't imagine logic inside the program that utilizes block count to pay rent automatically

**Review by NikVolf — CHANGES_REQUESTED** [review]
Syscall definitely should not take block count for paying, as completely non-deterministic

For extrinsic, this might be ok

File: pallets/gear/src/manager/journal.rs [hunk 3]
```diff
@@ -619,4 +619,46 @@ where
...
+        let value = Pallet::<T>::rent_fee_for(block_count);
+
+        ProgramStorageOf::<T>::update_active_program(program_id, |program| {
+            CurrencyOf::<T>::transfer(
+                &from,
+                &block_author,
+                value,
+                ExistenceRequirement::AllowDeath,
+            )
+            .unwrap_or_else(|e| unreachable!("Failed to transfer value: {:?}", e));
```
Comment (NikVolf) [inline pallets/gear/src/manager/journal.rs:639]: I'd avoid panics here and made `pay_rent` fallible

File: pallets/gear/src/lib.rs [hunk 4]
```diff
@@ -1677,6 +1688,59 @@ pub mod pallet {
...
+                        change: ProgramChangeKind::ExpirationChanged {
+                            expiration: expiration_block,
+                        },
+                    });
+
+                    Ok(())
+                },
+            )
+            .map_err(|e| {
+                log::debug!("update_active_program failed: {e:?}");
```
Comment (NikVolf) [inline pallets/gear/src/lib.rs:1738]: better to include `program_id` in log message

File: pallets/gear/src/manager/journal.rs (hunk 3 above)
Comment (gshep) [reply pallets/gear/src/manager/journal.rs:643]: Agree that panics should be avoided. However value cannot be transferred directly within syscall but used to reduce ValueCounter which is set to the actual free balance of the program before execution. If ValueCaounter cannot be updated the `pay_rent` fails with an error.

So this call in JournalNote processing should not fail ever.

File: pallets/gear/src/manager/journal.rs [hunk 5]
```diff
@@ -619,4 +619,46 @@ where
...
+            Pallet::<T>::deposit_event(Event::ProgramChanged {
+                id: program_id,
+                change: ProgramChangeKind::ExpirationChanged {
+                    expiration: expiration_block,
+                },
+            });
+        })
+        .unwrap_or_else(|e| {
+            log::debug!("Hold period may only be set for active program {e:?}");
+        });
```
[the remark this answers is not in the document]
Comment (gshep) [reply pallets/gear/src/manager/journal.rs]: ok

File: pallets/gear/src/lib.rs (hunk 4 above)
Comment (gshep) [reply pallets/gear/src/lib.rs]: ok

File: core-processor/src/ext.rs [hunk 6]
```diff
@@ -670,6 +674,46 @@ impl EnvExt for Ext {
...
+        let block_count = u32::try_from(rent / self.context.rent_cost).unwrap_or(u32::MAX);
+
+        let old_paid_blocks = self
+            .context
+            .program_rents
+            .get(&program_id)
+            .copied()
+            .unwrap_or(0);
+
+        let (paid_blocks, blocks_to_pay) = match old_paid_blocks.overflowing_add(block_count) {
```
Comment (breathx) [inline core-processor/src/ext.rs:687]: I'd also return error on overflow of u32
for case of handling such thing on pallet in overflow case I'd refund programs balance

File: core-processor/src/ext.rs [hunk 7]
```diff
@@ -670,6 +674,46 @@ impl EnvExt for Ext {
...
+            (count, false) => (count, block_count),
+            (_, true) => (u32::MAX, u32::MAX - old_paid_blocks),
+        };
+
+        if blocks_to_pay == 0 {
+            return Ok(rent);
+        }
+
+        let cost = self.context.rent_cost * u128::from(blocks_to_pay);
```
Comment (breathx) [inline core-processor/src/ext.rs:696]: unsafe thing

File: core-processor/src/ext.rs [hunk 8]
```diff
@@ -670,6 +674,46 @@ impl EnvExt for Ext {
...
+        match self.context.value_counter.reduce(cost) {
+            ChargeResult::Enough => {
+                self.context
+                    .program_rents
+                    .entry(program_id)
+                    .and_modify(|rent_blocks| *rent_blocks = paid_blocks)
+                    .or_insert(block_count);
+            }
+            ChargeResult::NotEnough => {
+                return Err(ExecutionError::NotEnoughValueForRent {
```
Comment (breathx) [inline core-processor/src/ext.rs:706]: better to add info about blocks amount (I would still go with approach of amount of blocks)

File: core-processor/src/processing.rs [hunk 9]
```diff
@@ -356,6 +359,16 @@ pub fn process_success(
         });
     }
 
+    // Must be handled after processing programs creation.
```
Comment (breathx) [inline core-processor/src/processing.rs:362]: It needs to be considered what to do if created program with inexistent code id (gr_create_program + gr_pay_rent)

Comment (breathx) [reply core-processor/src/processing.rs:362]: Maybe just refund on notes handling would work (provide test for it please)

File: gstd/src/exec.rs [hunk 10]
```diff
@@ -90,6 +90,22 @@ pub fn wake_delayed(message_id: MessageId, delay: u32) -> Result<()> {
     gcore::exec::wake_delayed(message_id.into(), delay).map_err(Into::into)
 }
 
+/// Pay specified rent for the program.
```
Comment (breathx) [inline gstd/src/exec.rs:93]: Specify anywhere return type

File: pallets/gear/src/benchmarking/mod.rs [hunk 11]
```diff
@@ -419,6 +422,26 @@ benchmarks! {
         assert!(MailboxOf::<T>::is_empty(&caller));
     }
 
+    pay_rent {
```
Comment (breathx) [inline pallets/gear/src/benchmarking/mod.rs:425]: handle default free period here for verification please

File: pallets/gear/src/lib.rs [hunk 12]
```diff
@@ -1677,6 +1688,59 @@ pub mod pallet {
...
+            program_id: ProgramId,
+            block_count: BlockNumberFor<T>,
+        ) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+
+            ProgramStorageOf::<T>::update_active_program(
+                program_id,
+                |program| -> Result<(), Error<T>> {
+                    let block_author =
+                        Authorship::<T>::author().ok_or(Error::<T>::BlockAuthorNotFound)?;
```
Comment (breathx) [inline pallets/gear/src/lib.rs:1706]: This error should be unreachable, isn't it?

File: pallets/gear/src/lib.rs [hunk 13]
```diff
@@ -1230,6 +1235,12 @@ pub mod pallet {
         pub fn run_call() -> Call<T> {
             Call::run {}
         }
+
+        pub fn rent_fee_for(block_count: BlockNumberFor<T>) -> BalanceOf<T> {
+            let block_count: u64 = block_count.unique_saturated_into();
+
+            RentCostPerBlockOf::<T>::get() * block_count.unique_saturated_into()
```
Comment (breathx) [inline pallets/gear/src/lib.rs:1242]: overflow

File: pallets/gear/src/lib.rs [hunk 14]
```diff
@@ -1677,6 +1688,59 @@ pub mod pallet {
...
+                    CurrencyOf::<T>::transfer(
+                        &who,
+                        &block_author,
+                        Self::rent_fee_for(block_count),
+                        ExistenceRequirement::AllowDeath,
+                    )
+                    .map_err(|_| Error::<T>::InsufficientBalanceForReserve)?;
+
+                    let old_block = program.expiration_block;
+                    let expiration_block = old_block.saturating_add(block_count);
```
Comment (breathx) [inline pallets/gear/src/lib.rs:1717]: overflow should be handled

File: pallets/gear/src/lib.rs [hunk 15]
```diff
@@ -1677,6 +1688,59 @@ pub mod pallet {
...
+                    program.expiration_block = expiration_block;
+
+                    let task = ScheduledTask::PauseProgram(program_id);
+                    let r = TaskPoolOf::<T>::delete(old_block, task.clone());
```
Comment (breathx) [inline pallets/gear/src/lib.rs:1721]: return errors?

File: pallets/gear/src/lib.rs [hunk 16]
```diff
@@ -1677,6 +1688,59 @@ pub mod pallet {
...
+                    log::debug!("TaskPool::delete result = {r:?}");
+
+                    let r = TaskPoolOf::<T>::add(expiration_block, task);
```
Comment (breathx) [inline pallets/gear/src/lib.rs:1724]: and here

File: pallets/gear/src/lib.rs [hunk 17]
```diff
@@ -1677,6 +1688,59 @@ pub mod pallet {
...
+                        change: ProgramChangeKind::ExpirationChanged {
+                            expiration: expiration_block,
+                        },
+                    });
+
+                    Ok(())
+                },
+            )
+            .map_err(|e| {
+                log::debug!("update_active_program {program_id} failed: {e:?}");
```
Comment (breathx) [inline pallets/gear/src/lib.rs:1738]: Expand debugging

File: pallets/gear/src/manager/journal.rs (hunk 3 above)
Comment (breathx) [reply pallets/gear/src/manager/journal.rs:643]: Yep, panicking here is totally pre-checked, but check my previous comments related to different kind of overflows and saturating operations.
E.g. program payed till u32::MAX / 2 and in case of payment for u32::MAX more we will lose u32::MAX - u32::MAX/2 funds, cause it already transferred above, but bn didn't increased properly.

File: pallets/gear/src/tests.rs [hunk 18]
```diff
@@ -1281,6 +1282,11 @@ fn read_state_bn_and_timestamp_works() {
         ));
 
         let program_id = utils::get_last_program_id();
+        assert_ok!(Gear::pay_rent(
```
Comment (breathx) [inline pallets/gear/src/tests.rs:1285]: please, avoid such repetitions in tests by increasing free period for tests. to tests specific logic of rents, hack the stuff in tests directly by increasing current/start block number

File: pallets/gear/src/tests.rs [hunk 19]
```diff
@@ -5403,6 +5411,63 @@ fn uninitialized_program_terminates_on_pause() {
     });
 }
 
+#[test]
+fn test_pay_rent_syscall_works() {
```
Comment (breathx) [inline pallets/gear/src/tests.rs:5415]: test for extrinsic itself?

File: pallets/gear/src/tests.rs [hunk 20]
```diff
@@ -5403,6 +5411,63 @@ fn uninitialized_program_terminates_on_pause() {
...
+            RuntimeOrigin::signed(USER_2),
+            pay_rent_id,
+            test_syscalls::Kind::PayRent(pay_rent_id.into_origin().into(), rent).encode(),
+            20_000_000_000,
+            0,
+        ));
+
+        run_to_next_block(None);
+
+        let program = ProgramStorageOf::<Test>::get_program(pay_rent_id)
```
Comment (breathx) [inline pallets/gear/src/tests.rs:5448]: test for corner cases with balances for program?

**Review by breathx — CHANGES_REQUESTED** [review]
@NikVolf why don't you want to go with bn in gr_pay_rent?

Someday "free" (const) syscalls would be applied and we could easily throw costs in user space so programs could calculate it manually. How to handle case when program requires exactly N blocks to pay? we could at least return it in syscall result as well.

**Comment by breathx** [comment]
Maybe we should consider somehow case when somebody payed for uninitialised program and program fails its initialisation? (may be considered in future, better to open task for now)

**Comment by breathx** [comment]
Furthermore, we should adjust payment amount! and somewhere validate that its greater then existential deposit (as this one used atm)
For current values we have:
payment for u32::MAX amount of blocks (4_294_967_295) will be: 4294967295 * 500 / 10**12 = 2.1474836475 TOKENS, thats too low value to pay for 45 years of program storing with 3s block time xd

I offer something like ~200 TOKENS per month: 
per_block_cost = 250_000_000 value, then with 3sec block time it will be 216 tokens per month

File: pallets/gear/src/lib.rs (hunk 12 above)
Comment (gshep) [reply pallets/gear/src/lib.rs:1706]: It maybe however it is easier to just return an error from the extrinsic

File: core-processor/src/ext.rs [hunk 21]
```diff
@@ -670,6 +674,46 @@ impl EnvExt for Ext {
         Ok(self.context.message_context.current().id())
     }
 
+    fn pay_rent(&mut self, program_id: ProgramId, rent: u128) -> Result<u128, Self::Error> {
+        let block_count = u32::try_from(rent / self.context.rent_cost).unwrap_or(u32::MAX);
```
[the remark this answers is not in the document]
Comment (gshep) [reply core-processor/src/ext.rs:678]: `if rent_cost.is_zero() { return Ok(()) }`

File: core-processor/src/ext.rs (hunk 8 above)
Comment (gshep) [reply core-processor/src/ext.rs:702]: discussed: return the `rent`

File: gsys/src/lib.rs [hunk 22]
```diff
@@ -244,6 +233,33 @@ where
     }
 }
 
+/// Represents type defining concatenated hash with value. 20 bytes.
+#[repr(C, packed)]
+#[derive(Default)]
+pub struct LengthWithValue {
+    pub length: Length,
+    pub value: Value,
+}
```
Comment (gshep) [inline gsys/src/lib.rs:242]: plus paid block count

File: pallets/gear/src/internal.rs [hunk 23]
```diff
@@ -978,4 +981,40 @@ where
...
+    }
+
+    pub(crate) fn update_expiration_block(
+        program_id: ProgramId,
+        program: &mut ActiveProgram<BlockNumberFor<T>>,
+        new_expiration_block: BlockNumberFor<T>,
+    ) {
+        let task = ScheduledTask::PauseProgram(program_id);
+        let r = TaskPoolOf::<T>::delete(program.expiration_block, task.clone());
+        log::debug!("TaskPool::delete result = {r:?}");
```
Comment (gshep) [inline pallets/gear/src/internal.rs:1006]: `unwarp_or_else unreachable`

**Comment by breathx** [comment]
add different benches for extrinsic

File: core-processor/src/processing.rs (hunk 9 above)
Comment (gshep) [reply core-processor/src/processing.rs:362]: discussed

File: pallets/gear/src/lib.rs (hunk 12 above)
Comment (gshep) [reply pallets/gear/src/lib.rs:1706]: `unwrap_or_else unreachable`

Comment (gshep) [reply pallets/gear/src/lib.rs:1706]: plus extract function

File: pallets/gear/src/lib.rs (hunk 17 above)
Comment (gshep) [reply pallets/gear/src/lib.rs:1728]: Rephrase the message

**Comment by gshep** [comment]
> 216

discussed to set to `125_000_000`

File: core-processor/src/ext.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (ark0f) [reply core-processor/src/ext.rs:678]: I don't think we will have zero rent cost

**Review by ark0f — COMMENTED** [review]
Awaiting for additional changes from conversations

File: core-processor/src/common.rs [hunk 24]
```diff
@@ -409,6 +421,8 @@ pub trait JournalHandler {
         destination: ProgramId,
         err: SimpleSignalError,
     );
+    /// Pay rent for the program.
+    fn pay_rent(&mut self, payer: ProgramId, program_id: ProgramId, block_count: u32);
```
[the remark this answers is not in the document]
Comment (gshep) [reply core-processor/src/common.rs:425]: ok

File: gsys/src/lib.rs [hunk 25]
```diff
@@ -233,26 +233,30 @@ where
     }
 }
 
-/// Represents type defining concatenated hash with value. 20 bytes.
+/// Represents type defining concatenated block number and value with length. 24 bytes.
 #[repr(C, packed)]
 #[derive(Default)]
-pub struct LengthWithValue {
+pub struct LengthWithBlockNumberValue {
```
Comment (ark0f) [inline gsys/src/lib.rs:239]:
```suggestion
pub struct LengthWithBlockNumberAndValue {
```

File: runtime/gear/src/constants.rs [hunk 26]
```diff
@@ -23,6 +23,9 @@ pub mod currency {
     /// The existential deposit.
     pub const EXISTENTIAL_DEPOSIT: Balance = 500;
 
+    /// The program rent cost per block.
+    pub const COST_PER_BLOCK: Balance = 125_000_000;
```
Comment (ark0f) [inline runtime/gear/src/constants.rs:27]:
```suggestion
    pub const RENT_COST_PER_BLOCK: Balance = 125_000_000;
```

**Review by ark0f — APPROVED** [review]

File: examples/binaries/sys-calls/src/lib.rs [hunk 27]
```diff
@@ -236,6 +238,9 @@ mod wasm {
                 let actual_mid: [u8; 32] = msg::id().into();
                 assert_eq!(expected_mid, actual_mid, "Kind::MessageId: mid test failed");
             }
+            Kind::PayProgramRent(program_id, rent) => {
+                exec::pay_program_rent(program_id.into(), rent).expect("Not enough value");
```
Comment (breathx) [inline examples/binaries/sys-calls/src/lib.rs:242]: assert return result here?

File: gstd/src/exec.rs [hunk 28]
```diff
@@ -90,6 +90,24 @@ pub fn wake_delayed(message_id: MessageId, delay: u32) -> Result<()> {
...
+/// rent value and the count of paid blocks.
+///
+/// # Examples
+///
+/// ```
+/// use gstd::{exec, ActorId};
+///
+/// #[no_mangle]
+/// extern "C" fn handle() {
+///     let (remaining_rent, paid_block_count) =
```
Comment (breathx) [inline gstd/src/exec.rs:103]:
```suggestion
///     let (unused_value, paid_block_count) =
```

File: gsys/src/lib.rs [hunk 29]
```diff
@@ -392,6 +412,16 @@ extern "C" {
     /// - `program_id`: `const ptr` for program id.
     pub fn gr_origin(program_id: *mut Hash);
 
+    /// Fallible `gr_pay_program_rent` syscall.
+    ///
+    /// Arguments type:
+    /// - `rent_pid`: `const ptr` for program id and rent value.
+    /// - `err_bn_value`: `mut ptr` for concatenated error length, paid block count and remaining rent value.
```
Comment (breathx) [inline gsys/src/lib.rs:419]:
```suggestion
    /// - `err_bn_value`: `mut ptr` for concatenated error length, paid block count and unused rent value.
```

Comment (breathx) [reply gsys/src/lib.rs:419]: remaining in this context seems for me like amount I have to extra pay

File: pallets/gear/src/lib.rs [hunk 30]
```diff
@@ -1669,6 +1678,42 @@ pub mod pallet {
...
+        #[pallet::call_index(8)]
+        #[pallet::weight(<T as Config>::WeightInfo::pay_program_rent())]
+        pub fn pay_program_rent(
+            origin: OriginFor<T>,
+            program_id: ProgramId,
+            block_count: BlockNumberFor<T>,
+        ) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+
+            let block_author = Authorship::<T>::author()
```
Comment (breathx) [inline pallets/gear/src/lib.rs:1692]: block author query may be included in pay rent impl

File: pallets/gear/src/internal.rs [hunk 31]
```diff
@@ -978,4 +984,51 @@ where
...
+
+        CurrencyOf::<T>::transfer(
+            from,
+            block_author,
+            Self::rent_fee_for(blocks_to_pay),
+            ExistenceRequirement::AllowDeath,
+        )?;
+
+        let task = ScheduledTask::PauseProgram(program_id);
+        TaskPoolOf::<T>::delete(program.expiration_block, task.clone())
```
Comment (breathx) [inline pallets/gear/src/internal.rs:1017]:
```suggestion
        TaskPoolOf::<T>::delete(old_expiration_block, task.clone())
```

File: pallets/gear/src/tests.rs [hunk 32]
```diff
@@ -5404,6 +5408,176 @@ fn uninitialized_program_terminates_on_pause() {
...
+        );
+        assert!(TaskPoolOf::<Test>::contains(
+            &program.expiration_block,
+            &ScheduledTask::PauseProgram(pay_rent_id)
+        ));
+    });
+}
+
+#[test]
+fn pay_program_rent_extrinsic_works() {
```
Comment (breathx) [inline pallets/gear/src/tests.rs:5516]: case of pay more for u32::MAX (refund) case?

File: pallets/gear/src/tests.rs [hunk 33]
```diff
@@ -5404,6 +5408,176 @@ fn uninitialized_program_terminates_on_pause() {
...
+            20_000_000_000,
+            0,
+        ));
+
+        run_to_next_block(None);
+
+        assert_eq!(balance_before, Balances::free_balance(pay_rent_account_id));
+
+        // try to pay greater rent than available value
+        assert_ok!(Gear::send_message(
```
Comment (breathx) [inline pallets/gear/src/tests.rs:5471]: assert message panicked

**Review by breathx — CHANGES_REQUESTED** [review]
Finish line!

File: pallets/gear/src/tests.rs (hunk 32 above)
Comment (gshep) [reply pallets/gear/src/tests.rs:5516]: frankly speaking this is a bit tricky since `BlockNumber = u64` for Test runtime )

File: pallets/gear/src/tests.rs [hunk 34]
```diff
@@ -5404,6 +5408,176 @@ fn uninitialized_program_terminates_on_pause() {
...
+            Kind::PayProgramRent([0u8; 32], rent).encode(),
+            20_000_000_000,
+            0,
+        ));
+
+        run_to_next_block(None);
+
+        assert_eq!(balance_before, Balances::free_balance(pay_rent_account_id));
+
+        // try to pay greater rent than available value
```
[the remark this answers is not in the document]
Comment (gshep) [reply pallets/gear/src/tests.rs:5470]: Nope, at this line `program_value` is greater the actual balance since the program has paid some rent before

File: pallets/gear/src/tests.rs [hunk 35]
```diff
@@ -5404,6 +5408,176 @@ fn uninitialized_program_terminates_on_pause() {
...
+            .expect("program should exist");
+        assert_eq!(
+            expiration_block.saturating_add(block_count),
+            program.expiration_block
+        );
+        assert!(TaskPoolOf::<Test>::contains(
+            &program.expiration_block,
+            &ScheduledTask::PauseProgram(pay_rent_id)
+        ));
+    });
```
[the remark this answers is not in the document]
Comment (gshep) [reply pallets/gear/src/tests.rs:5512]: frankly speaking it is already there however I will adjust it a bit.

File: pallets/gear/src/tests.rs [hunk 36]
```diff
@@ -5404,6 +5408,176 @@ fn uninitialized_program_terminates_on_pause() {
...
+
+        run_to_next_block(None);
+
+        assert_eq!(balance_before, Balances::free_balance(pay_rent_account_id));
+        let program = ProgramStorageOf::<Test>::get_program(pay_rent_id)
+            .and_then(|p| ActiveProgram::try_from(p).ok())
+            .expect("program should exist");
+        assert_eq!(expiration_block, program.expiration_block);
+
+        // pay maximum possible rent
```
[the remark this answers is not in the document]
Comment (gshep) [reply pallets/gear/src/tests.rs:5487]: The message processing should not fail in this case. It is about refunding actually

File: pallets/gear/src/tests.rs [hunk 37]
```diff
@@ -5486,29 +5486,83 @@ fn pay_program_rent_syscall_works() {
...
+        assert_failed(
+            message_id,
+            ActorExecutionErrorReason::Trap(TrapExplanation::Panic(error_text.into())),
+        );
+
         // pay maximum possible rent
-        let block_count: BlockNumberFor<Test> = u32::MAX.into();
+        let block_count = u32::MAX;
         let required_value = Gear::rent_fee_for(block_count - expiration_block);
         assert_ok!(Gear::send_message(
```
Comment (breathx) [inline pallets/gear/src/tests.rs:5567]: assert succeed

File: pallets/gear/src/tests.rs [hunk 38]
```diff
@@ -5486,29 +5486,83 @@ fn pay_program_rent_syscall_works() {
...
+            TrapExplanation::Ext(ExtError::Execution(ExecutionError::MaximumBlockCountPaid))
+        );
+        assert_failed(
+            message_id,
+            ActorExecutionErrorReason::Trap(TrapExplanation::Panic(error_text.into())),
+        );
+
         // pay maximum possible rent
-        let block_count: BlockNumberFor<Test> = u32::MAX.into();
+        let block_count = u32::MAX;
```
Comment (breathx) [inline pallets/gear/src/tests.rs:5565]: expiration block != block count

File: pallets/gear/src/tests.rs [hunk 39]
```diff
@@ -5520,10 +5574,7 @@ fn pay_program_rent_syscall_works() {
         let program = ProgramStorageOf::<Test>::get_program(pay_rent_id)
```
Comment (breathx) [inline pallets/gear/src/tests.rs:5583]: add comment why balance hasn't been changed

**Review by breathx — APPROVED** [review]
👍🏻

**Review by NikVolf — APPROVED** [review]

**Review by NikVolf — APPROVED** [review]
